cpp/src/community/k_truss_impl.cuh

```diff
@@ -214,121 +409,501 @@ k_truss(raft::handle_t const& handle,
-    cugraph::fill_edge_property(handle, cur_graph_view, edge_mask.mutable_view(), bool{true});
+    size_t prev_chunk_size = 0;  // FIXME: Add support for chunking
 
     while (true) {
-      // FIXME: This approach is very expensive when invalidating only few edges per iteration
-      // and should be address.
-      auto edge_triangle_counts =
-        edge_triangle_count<vertex_t, edge_t, multi_gpu>(handle, cur_graph_view);
+      cur_graph_view.clear_edge_mask();
+      cur_graph_view.attach_edge_mask(dodg_mask.view());
 
-      // Mask all the edges that have k - 2 count
+      // Extract weak edges
+      auto [weak_edgelist_srcs, weak_edgelist_dsts] =
+        extract_transform_e(handle,
+                            cur_graph_view,
+                            edge_src_dummy_property_t{}.view(),
+                            edge_dst_dummy_property_t{}.view(),
+                            edge_triangle_counts.view(),
+                            extract_weak_edges<vertex_t, edge_t>{k});
+
+      auto weak_edgelist_first =
+        thrust::make_zip_iterator(weak_edgelist_srcs.begin(), weak_edgelist_dsts.begin());
+      auto weak_edgelist_last =
+        thrust::make_zip_iterator(weak_edgelist_srcs.end(), weak_edgelist_dsts.end());
+
+      // Perform nbr_intersection of the weak edges from the undirected
+      // graph view
+      cur_graph_view.clear_edge_mask();
+
+      // Attach the weak edge mask
+      cur_graph_view.attach_edge_mask(weak_edges_mask.view());
+
+      auto [intersection_offsets, intersection_indices] = per_v_pair_dst_nbr_intersection(
+        handle, cur_graph_view, weak_edgelist_first, weak_edgelist_last, false);
+
+      // Identify (p, q) edges, and form edges (p, q), (p, r) and (q, r)
+      // To avoid overcompensation, redirect all edges in the triangle to follow this unique
+      // pattern: (p, q) then (q, r) then (p, r)
+
+      auto triangles_from_weak_edges = allocate_dataframe_buffer<
+        thrust::tuple<vertex_t, vertex_t, vertex_t, vertex_t, vertex_t, vertex_t>>(
+        intersection_indices.size(), handle.get_stream());
+
+      // Form (p, q) edges
+      // Extract triangle from weak
+      thrust::tabulate(
+        handle.get_thrust_policy(),
+        get_dataframe_buffer_begin(triangles_from_weak_edges),
+        get_dataframe_buffer_end(triangles_from_weak_edges),
+        extract_triangles_from_weak_edges<vertex_t, edge_t>{
+          prev_chunk_size,
+          raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()),
+          raft::device_span<vertex_t const>(intersection_indices.data(),
+                                            intersection_indices.size()),
+          raft::device_span<vertex_t const>(weak_edgelist_srcs.data(), weak_edgelist_srcs.size()),
+          raft::device_span<vertex_t const>(weak_edgelist_dsts.data(), weak_edgelist_dsts.size())});
```

Comment:
Instead of storing all three edges in each triangle, what happens 

1) if we store only three triangle end points (a, b, c where out-degree(a) < out-degree(b) < out-degree(c) using the same tie breaking mechanism to create a DODG graph),
2) Shuffle based on edge partitioning using (a, b)
3) Remove duplicates
4) Re-create an edge list (a, b) (a, c), (b, c) from each (a, b, c)
5) Shuffle the result edge list
6) Decrease edge counts.

Won't this work? This sounds simpler & cheaper.

Reply:
And we haven't implemented truss decomposition, right?

K-Truss should be implemented as truss decomposition with min_k = max_k = k